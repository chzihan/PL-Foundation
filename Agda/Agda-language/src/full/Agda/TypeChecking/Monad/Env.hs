
module Agda.TypeChecking.Monad.Env where

import Control.Monad.Reader
import qualified Data.List as List
import Data.Monoid

import Agda.Syntax.Common
import Agda.Syntax.Abstract.Name

import Agda.TypeChecking.Monad.Base

-- | Get the name of the current module, if any.
{-# SPECIALIZE currentModule :: TCM ModuleName #-}
{-# SPECIALIZE currentModule :: ReduceM ModuleName #-}
currentModule :: MonadReader TCEnv m => m ModuleName
currentModule = asks envCurrentModule

-- | Set the name of the current module.
withCurrentModule :: ModuleName -> TCM a -> TCM a
withCurrentModule m =
    local $ \e -> e { envCurrentModule = m }

-- | Get the number of variables bound by anonymous modules.
{-# SPECIALIZE getAnonymousVariables :: ModuleName -> TCM Nat #-}
{-# SPECIALIZE getAnonymousVariables :: ModuleName -> ReduceM Nat #-}
getAnonymousVariables :: MonadReader TCEnv m => ModuleName -> m Nat
getAnonymousVariables m = do
  ms <- asks envAnonymousModules
  return $ sum [ n | (m', n) <- ms, mnameToList m' `List.isPrefixOf` mnameToList m ]

-- | Add variables bound by an anonymous module.
withAnonymousModule :: ModuleName -> Nat -> TCM a -> TCM a
withAnonymousModule m n =
  local $ \e -> e { envAnonymousModules   = (m, n) : envAnonymousModules e
                  }

-- | Set the current environment to the given
withEnv :: TCEnv -> TCM a -> TCM a
withEnv env = local $ \ env0 -> env
  -- Keep persistent settings
  { envAllowDestructiveUpdate = envAllowDestructiveUpdate env0
  , envPrintMetasBare         = envPrintMetasBare env0
  }

-- | Get the current environment
getEnv :: TCM TCEnv
getEnv = ask

-- | Increases the module nesting level by one in the given
-- computation.
withIncreasedModuleNestingLevel :: TCM a -> TCM a
withIncreasedModuleNestingLevel =
  local (\e -> e { envModuleNestingLevel =
                     envModuleNestingLevel e + 1 })

-- | Set highlighting level
withHighlightingLevel :: HighlightingLevel -> TCM a -> TCM a
withHighlightingLevel h = local $ \e -> e { envHighlightingLevel = h }

-- | Restore setting for 'ExpandLast' to default.
doExpandLast :: TCM a -> TCM a
doExpandLast = local $ \ e -> e { envExpandLast = ExpandLast }

dontExpandLast :: TCM a -> TCM a
dontExpandLast = local $ \ e -> e { envExpandLast = DontExpandLast }

-- | If the reduced did a proper match (constructor or literal pattern),
--   then record this as simplification step.
{-# SPECIALIZE performedSimplification :: TCM a -> TCM a #-}
performedSimplification :: MonadReader TCEnv m => m a -> m a
performedSimplification = local $ \ e -> e { envSimplification = YesSimplification }

{-# SPECIALIZE performedSimplification' :: Simplification -> TCM a -> TCM a #-}
performedSimplification' :: MonadReader TCEnv m => Simplification -> m a -> m a
performedSimplification' simpl = local $ \ e -> e { envSimplification = simpl `mappend` envSimplification e }

getSimplification :: MonadReader TCEnv m => m Simplification
getSimplification = asks envSimplification

-- * Controlling reduction.

-- | Lens for 'AllowedReductions'.
updateAllowedReductions :: (AllowedReductions -> AllowedReductions) -> TCEnv -> TCEnv
updateAllowedReductions f e = e { envAllowedReductions = f (envAllowedReductions e) }

modifyAllowedReductions :: (AllowedReductions -> AllowedReductions) -> TCM a -> TCM a
modifyAllowedReductions = local . updateAllowedReductions

putAllowedReductions :: AllowedReductions -> TCM a -> TCM a
putAllowedReductions = modifyAllowedReductions . const

-- | Reduce @Def f vs@ only if @f@ is a projection.
onlyReduceProjections :: TCM a -> TCM a
onlyReduceProjections = putAllowedReductions [ProjectionReductions]

-- | Allow all reductions except for non-terminating functions (default).
allowAllReductions :: TCM a -> TCM a
allowAllReductions = putAllowedReductions allReductions

-- | Allow all reductions including non-terminating functions.
allowNonTerminatingReductions :: TCM a -> TCM a
allowNonTerminatingReductions = putAllowedReductions $ [NonTerminatingReductions] ++ allReductions

-- * Concerning 'envInsideDotPattern'

insideDotPattern :: TCM a -> TCM a
insideDotPattern = local $ \e -> e { envInsideDotPattern = True }

isInsideDotPattern :: TCM Bool
isInsideDotPattern = asks envInsideDotPattern

-- | Don't use call-by-need evaluation for the given computation.
callByName :: TCM a -> TCM a
callByName = local $ \ e -> e { envCallByNeed = False }

