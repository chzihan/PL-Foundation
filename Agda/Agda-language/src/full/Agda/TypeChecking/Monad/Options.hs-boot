module Agda.TypeChecking.Monad.Options where

import Control.Monad.Reader

import Agda.TypeChecking.Monad.Base
import Agda.Utils.FileName

getIncludeDirs :: TCM [AbsolutePath]

type VerboseKey = String

hasVerbosity :: HasOptions m => VerboseKey -> Int -> m Bool
verboseS :: (MonadReader TCEnv m, HasOptions m) => VerboseKey -> Int -> m () -> m ()
