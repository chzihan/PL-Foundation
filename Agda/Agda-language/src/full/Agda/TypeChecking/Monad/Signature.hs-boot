
module Agda.TypeChecking.Monad.Signature where

import Agda.Syntax.Internal (ModuleName, Telescope)
import Agda.TypeChecking.Monad.Base (TCM, ReadTCState)

inFreshModuleIfFreeParams :: TCM a -> TCM a
lookupSection :: (Functor m, ReadTCState m) => ModuleName -> m Telescope

