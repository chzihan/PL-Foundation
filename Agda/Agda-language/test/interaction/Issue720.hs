{-# LANGUAGE RecordWildCards #-}

import Control.Monad
import System.Directory

import RunAgda

file          = "Issue720.agda"
interfaceFile = file ++ "i"

main :: IO ()
main = runAgda [] $ \(AgdaCommands { .. }) -> do

  let load = do
        send $ command "load" file (Just "Interactive Direct") (Just [])
        echoUntilPrompt

  -- Discard the first prompt.
  echoUntilPrompt

  -- Load the file, and wait for Agda to complete.
  load

  -- Remove the interface file.
  exists <- doesFileExist interfaceFile
  when exists $ removeFile interfaceFile

  -- Reload.
  load

  return ()
