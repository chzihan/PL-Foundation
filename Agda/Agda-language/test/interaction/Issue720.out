(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue720 (Issue720.agda). " t)
(agda2-highlight-add-annotations '(1 7 (keyword) nil) '(17 22 (keyword) nil))
(agda2-highlight-add-annotations '(1 7 (keyword) nil) '(8 16 (module) nil ("Issue720.agda" . 1)) '(17 22 (keyword) nil))
(agda2-highlight-add-annotations '(1 7 (keyword) nil) '(8 16 (module) nil ("Issue720.agda" . 1)) '(17 22 (keyword) nil))
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
(agda2-status-action "")
(agda2-info-action "*Type-checking*" "" nil)
(agda2-highlight-clear)
(agda2-info-action "*Type-checking*" "Checking Issue720 (Issue720.agda). " t)
(agda2-highlight-add-annotations '(1 7 (keyword) nil) '(17 22 (keyword) nil))
(agda2-highlight-add-annotations '(1 7 (keyword) nil) '(8 16 (module) nil ("Issue720.agda" . 1)) '(17 22 (keyword) nil))
(agda2-highlight-add-annotations '(1 7 (keyword) nil) '(8 16 (module) nil ("Issue720.agda" . 1)) '(17 22 (keyword) nil))
(agda2-status-action "Checked")
(agda2-info-action "*All Done*" "" nil)
((last . 1) . (agda2-goals-action '()))
